File: socialview-core/src/main/java/com/hendraanggrian/socialview/SociableView.java
package com.hendraanggrian.socialview;

import android.support.annotation.AttrRes;
import android.support.annotation.ColorInt;
import android.support.annotation.ColorRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import java.util.Collection;

/**
 * Base methods of all socialview's widgets.
 * The logic, however, are calculated in {@link SociableViewImpl} while the widgets are just
 * passing these methods to the attacher.
 *
 * @author Hendra Anggian (hendraanggrian@gmail.com)
 * @see SociableViewImpl
 */
public interface SociableView {

    boolean isHashtagEnabled();

    boolean isMentionEnabled();

    boolean isHyperlinkEnabled();

    void setHashtagEnabled(boolean enabled);

    void setMentionEnabled(boolean enabled);

    void setHyperlinkEnabled(boolean enabled);

    @ColorInt
    int getHashtagColor();

    @ColorInt
    int getMentionColor();

    @ColorInt
    int getHyperlinkColor();

    void setHashtagColor(@ColorInt int color);

    void setMentionColor(@ColorInt int color);

    void setHyperlinkColor(@ColorInt int color);

    void setHashtagColorRes(@ColorRes int colorRes);

    void setMentionColorRes(@ColorRes int colorRes);

    void setHyperlinkColorRes(@ColorRes int colorRes);

    void setHashtagColorAttr(@AttrRes int colorAttr);

    void setMentionColorAttr(@AttrRes int colorAttr);

    void setHyperlinkColorAttr(@AttrRes int colorAttr);

    void setOnHashtagClickListener(@Nullable OnSocialClickListener listener);

    void setOnMentionClickListener(@Nullable OnSocialClickListener listener);

    void setHashtagTextChangedListener(@Nullable SocialTextWatcher watcher);

    void setMentionTextChangedListener(@Nullable SocialTextWatcher watcher);

    @NonNull
    Collection<String> getHashtags();

    @NonNull
    Collection<String> getMentions();

    @NonNull
    Collection<String> getHyperlinks();
}

File: socialview-core/src/com/hendraanggrian/socialview/SociableView.kt
package com.hendraanggrian.socialview

import android.content.Context
import android.support.annotation.AttrRes
import android.support.annotation.ColorRes
import com.hendraanggrian.support.utils.content.getColor
import com.hendraanggrian.support.utils.content.getColor2

/**
 * Base methods of all socialview's widgets.
 * The logic, however, are calculated in [SociableViewImpl] while the widgets are just
 * passing these methods to the attacher.

 * @author Hendra Anggian (hendraanggrian@gmail.com)
 * *
 * @see SociableViewImpl
 */
interface SociableView {

    val internalContext: Context

    val hashtags: Collection<String>
    val mentions: Collection<String>
    val hyperlinks: Collection<String>

    var isHashtagEnabled: Boolean
    var isMentionEnabled: Boolean
    var isHyperlinkEnabled: Boolean

    var hashtagColor: Int
    var mentionColor: Int
    var hyperlinkColor: Int

    fun setHashtagColorRes(@ColorRes id: Int) {
        hashtagColor = internalContext.getColor2(id)
    }

    fun setMentionColorRes(@ColorRes id: Int) {
        mentionColor = internalContext.getColor2(id)
    }

    fun setHyperlinkColorRes(@ColorRes id: Int) {
        hyperlinkColor = internalContext.getColor2(id)
    }

    fun setHashtagColorAttr(@AttrRes id: Int) {
        hashtagColor = internalContext.theme.getColor(id, true)
    }

    fun setMentionColorAttr(@AttrRes id: Int) {
        mentionColor = internalContext.theme.getColor(id, true)
    }

    fun setHyperlinkColorAttr(@AttrRes id: Int) {
        hyperlinkColor = internalContext.theme.getColor(id, true)
    }

    fun setOnHashtagClickListener(listener: ((SociableView, CharSequence) -> Unit)?)
    fun setOnMentionClickListener(listener: ((SociableView, CharSequence) -> Unit)?)

    fun setHashtagTextChangedListener(watcher: ((SociableView, CharSequence) -> Unit)?)
    fun setMentionTextChangedListener(watcher: ((SociableView, CharSequence) -> Unit)?)
}